flink-queryable-state/flink-queryable-state-java/src/main/java/org/apache/flink/queryablestate/network/AbstractServerBase.java

```diff
@@ -96,90 +113,158 @@ protected AbstractServerBase(
-				.childOption(ChannelOption.WRITE_BUFFER_LOW_WATER_MARK, LOW_WATER_MARK);
+		this.bindAddress = Preconditions.checkNotNull(bindAddress);
+		this.numEventLoopThreads = numEventLoopThreads;
+		this.numQueryThreads = numQueryThreads;
+
+		this.bindPortRange = new HashSet<>();
+		while (bindPortIterator.hasNext()) {
+			int port = bindPortIterator.next();
+			Preconditions.checkArgument(port >= 0 && port <= 65535,
+					"Invalid port configuration. Port must be between 0 and 65535, but was " + port + ".");
+			bindPortRange.add(port);
+		}
 	}
 
 	/**
 	 * Creates a thread pool for the query execution.
-	 *
-	 * @param numQueryThreads Number of query threads.
 	 * @return Thread pool for query execution
 	 */
-	private ExecutorService createQueryExecutor(int numQueryThreads) {
+	private ExecutorService createQueryExecutor() {
 		ThreadFactory threadFactory = new ThreadFactoryBuilder()
 				.setDaemon(true)
 				.setNameFormat("Flink " + getServerName() + " Thread %d")
 				.build();
-
 		return Executors.newFixedThreadPool(numQueryThreads, threadFactory);
 	}
 
+	/**
+	 * Returns the thread-pool responsible for processing incoming requests.
+	 */
 	protected ExecutorService getQueryExecutor() {
 		return queryExecutor;
 	}
 
+	/**
+	 * Gets the name of the server. This is useful for debugging.
+	 * @return The name of the server.
+	 */
 	public String getServerName() {
 		return serverName;
 	}
 
+	/**
+	 * Returns the {@link AbstractServerHandler handler} to be used for
+	 * serving the incoming requests.
+	 */
 	public abstract AbstractServerHandler<REQ, RESP> initializeHandler();
 
 	/**
+	 * Returns the address of this server.
+	 *
+	 * @return AbstractServerBase address
+	 * @throws IllegalStateException If server has not been started yet
+	 */
+	public KvStateServerAddress getServerAddress() {
+		Preconditions.checkState(serverAddress != null, "Server " + serverName + " has not been started.");
+		return serverAddress;
+	}
+
+	/**
 	 * Starts the server by binding to the configured bind address (blocking).
 	 * @throws InterruptedException If interrupted during the bind operation
 	 */
 	public void start() throws InterruptedException {
 		Preconditions.checkState(serverAddress == null,
-				"Server " + serverName + " has already been started @ " + serverAddress + '.');
-
-		this.handler = initializeHandler();
-		bootstrap.childHandler(new ServerChannelInitializer<>(handler));
-
-		Channel channel = bootstrap.bind().sync().channel();
-		InetSocketAddress localAddress = (InetSocketAddress) channel.localAddress();
-		serverAddress = new KvStateServerAddress(localAddress.getAddress(), localAddress.getPort());
+				"Server " + serverName + " already running @ " + serverAddress + '.');
+
+		Iterator<Integer> portIterator = bindPortRange.iterator();
+		while (portIterator.hasNext() && serverAddress == null) {
+			final int port = portIterator.next();
+			try {
+				attemptToBind(port);
+			} catch (Exception e) {
```

Comment:
Is there a more specific exception that will be thrown if the port is taken? We could catch specifically that one in `attemptToBind()` and return true or false depending on whether binding was succesful.